include/hc.hpp

```diff
@@ -117,6 +451,188 @@ inline uint64_t get_tick_frequency() {
+                                                                                  const Kernel&);
+
+    #if __KALMAR_ACCELERATOR__ == 2 || __KALMAR_CPU__ == 2
+        template<typename Kernel, int N>
+        friend
+        completion_future launch_cpu_task_async(const std::shared_ptr<Kalmar::KalmarQueue>&,
+                                                const Kernel&,
+                                                const extent<N>&);
+    #endif
+
+    // non-tiled parallel_for_each
+    // generic version
+    template<int N, typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<N>&,
+                                        const Kernel&);
+
+    // 1D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<1>&,
+                                        const Kernel&);
+
+    // 2D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<2>&,
+                                        const Kernel&);
+
+    // 3D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 3D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 2D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<2>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 1D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<1>&,
+                                        const Kernel&);
+
+    static constexpr unsigned int max_async_ops = 5u; // TODO: retrieve directly from HSA runtime.
+
+    std::shared_ptr<Kalmar::KalmarQueue> queue_;
+
+    explicit
+    accelerator_view(std::shared_ptr<Kalmar::KalmarQueue> queue)
+        : queue_{std::move(queue)} {}
+
+    template<typename InputIterator>
+    completion_future create_blocking_marker_(InputIterator f,
+                                              InputIterator l,
+                                              memory_scope scope,
+                                              std::true_type) const
```

Comment:
It's really not clear that `true_type` and `false_type` is related to the `static_assert` below, epecially since they aren't mutually exclusive. I think using `enable_if` would be cleaner.

```cpp
template<typename InputIterator, typename std::enable_if<(
    std::is_same<typename std::iterator_traits<InputIterator>::value_type, completion_future>{}
), int>::type = 0> 
```

And I think a macro and template alias can help reduce the noise:

```cpp
#define HCC_REQUIRES(...) typename std::enable_if<(__VA_ARGS__), int>::type = 0

template<class Iterator>
using iterator_value = typename std::iterator_traits<Iterator>::value_type;

template<typename InputIterator, 
    HCC_REQUIRES(std::is_same<iterator_value<InputIterator>, completion_future>{})
>
```

Plus this puts the type requirements for the template parameter next to it. 

But, I don't know how everyone else thinks about it?